Iot progress indicator

rawa-resul opened the pull request:
Closes #1116 #1074

**Summary**

- The new design of the progress indicator based on @lukefirth design from [here](https://pages.github.ibm.com/Watson-IoT/IoT-Design-Site/components/progressindicator/overview/)

_Vertical view & RTL_
![Screenshot 2020-07-07 at 09 57 37](https://user-images.githubusercontent.com/61942902/86755358-7d88af00-c039-11ea-9a53-8d420ae07c8a.png)
![Screenshot 2020-07-07 at 09 58 12](https://user-images.githubusercontent.com/61942902/86755367-7feb0900-c039-11ea-8319-2756249609f9.png)

_Horizontal view & RTL_
![Screenshot 2020-07-07 at 09 57 52](https://user-images.githubusercontent.com/61942902/86755562-a446e580-c039-11ea-9bbb-c532ac769f33.png) ![Screenshot 2020-07-07 at 09 58 04](https://user-images.githubusercontent.com/61942902/86755575-a5781280-c039-11ea-9f7c-06d79be23c8e.png)

_Completed steps with checkmarks_
![Screenshot 2020-07-07 at 09 59 34](https://user-images.githubusercontent.com/61942902/86755756-c9d3ef00-c039-11ea-96a3-0e71c7fd9b7e.png)

_Integrated into PageWizard_
![Screenshot 2020-07-07 at 09 58 28](https://user-images.githubusercontent.com/61942902/86755818-d3f5ed80-c039-11ea-9b46-e55981adf793.png)
![Screenshot 2020-07-07 at 09 58 43](https://user-images.githubusercontent.com/61942902/86755831-d5271a80-c039-11ea-8ef5-9c3d2f5b1b62.png)


**Change List (commits, features, bugs, etc)**

- Created a new ProgressIndicator component using functional rather than class component with backwards compatability
- Created a new ProgressStep component that supports the new styles & svgs
- Supports a new data structure for list items that can incorporate child elements acting as sub elements
- Main steps now show numbers in the icons
- Disabled steps can not be accessed and will be skipped
- Error/Warning steps can be accessed
- Added new prop `clickable` that removes clicking property from the steps (disabling them) - by default to false
- Removed `Styled` functions completely
- Added, Removed, Changed various test cases around the new progress indicator to adhere to new functionality
- Added new prop in PageWizardStep: `subStep` to declare sub steps

**Improvement suggestions**
- I was not able to implement the truncate/ellipsis and tooltip successfully because of width issues in vertical/horizontal view of the labels - it would be great if someone could look into that
- The icons we use in carbon add an invisible border I could not remove. Ideally we dont want this border so that the line sticks directly to the icon
- Currently it does not support passing down of ProgressStep components as children. You would need the items array. 
- Needs a new Skeleton that matches the new design

**Acceptance Test (how to verify the PR)**

- Go to Storybook and click on ProgressIndicator
- Go through the various stories and check that everything works
- Additionally, test the props such as clickable, showlabels, stepWidth and RTL
- Do the same in PageWizard
- Additionally, check that in PageWizard with validation you cannot move to the next step in the progress indicator unless form condition is satisfied. 
- Do the same in TableDetailWizard

tay1orjones commented on src/components/PageWizard/StatefulPageWizard.jsx:
```diff
@@ -76,23 +76,48 @@ const StatefulPageWizard = ({
-  const prevStep = currentStepIndex > 0 ? steps[currentStepIndex - 1] : undefined;
+
+  // skip disabled steps when clicking next or prev
+  const getAvailableStep = previous => {
+    let idx = previous ? currentStepIndex - 1 : currentStepIndex + 1;
+    while (true) {
+      if (idx >= 0 && idx < steps.length) {
+        if (steps[idx].disabled === true) {
+          if (previous && idx !== 0) {
+            idx -= 1;
+          } else if (idx !== steps.length - 1) {
+            idx += 1;
+          }
+        } else {
+          return steps[idx];
+        }
+      } else {
+        return undefined;
+      }
+    }
+  };
+
+  const nextStep = getAvailableStep();
+  const prevStep = getAvailableStep(true);
```
This signature is a bit confusing. Maybe the index should be the param?

```
  const nextStep = getAvailableStepAtIndex(currentStepIndex + 1);
  const prevStep = getAvailableStepAtIndex(currentStepIndex - 1);
```

rawa-resul replied:
yeah you are right. I changed it by created two simpler/shorter functions for nextStep and prevStep

rawa-resul commented on src/components/PageWizard/PageWizard.jsx:
```diff
@@ -52,6 +52,8 @@ export const PageWizardPropTypes = {
   /** Content to render before footer buttons (on left side, in LTR) */
   beforeFooterContent: PropTypes.node,
+  /** Make the ProgressIndicator clickable */
+  clickable: PropTypes.bool,
```
Thanks Taylor, I just made this change. All `clickable` props should be `isClickable` now.

davidicus commented on src/components/PageWizard/PageWizard.jsx:
```diff
@@ -1,8 +1,8 @@
 import PropTypes from 'prop-types';
-import { ProgressIndicator, ProgressStep } from 'carbon-components-react';
-import classnames from 'classnames';
+import classNames from 'classnames';
```
Does this not trigger a lint warning. I think we have a lint rule to standardize these imports to only use `classnames`

tay1orjones replied:
```suggestion
import classnames from 'classnames';
```

davidicus commented on src/components/ProgressIndicator/_progress-indicator.scss:
```diff
@@ -1,82 +1,306 @@
+    height: 2px;
+    background-color: #e0e0e0;
+    top: 11px;
+    left: 24px;
```
we shouldn't have any px values

davidicus commented on src/components/ProgressIndicator/_progress-indicator.scss:
```diff
@@ -1,82 +1,306 @@
+    position: relative;
+    text-align: left;
+    width: 8rem;
+    height: 68px;
```
no pixel values

davidicus commented on src/components/ProgressIndicator/_progress-indicator.scss:
```diff
@@ -1,82 +1,306 @@
-    margin-right: $spacing-03;
+  .sub-step {
+    .icon {
+      margin-top: 4px;
```
can this be a standard carbon token?

rawa-resul replied:
I will change it to $spacing-02

davidicus commented on src/components/ProgressIndicator/_progress-indicator.scss:
```diff
@@ -1,82 +1,306 @@
-  .#{$prefix}--progress-line {
-    top: 0;
+  svg text {
+    font-size: 18px;
```
no pixel values

davidicus commented on src/components/ProgressIndicator/_progress-indicator.scss:
```diff
@@ -1,82 +1,306 @@
+
+  .step-current {
+    &:not(.step-invalid) {
+      .main-step {
```
Not sure if these are new classes but we should be using BEM syntax to create these 
```
.#{$prefix}--progress-line-step
.#{$prefix}--progress-line-step__current
.#{$prefix}--progress-line-step__invalid
etc. 
```

davidicus commented on src/components/ProgressIndicator/_progress-indicator.scss:
```diff
@@ -1,82 +1,306 @@
+  .step-current {
+    &:not(.step-invalid) {
+      .main-step {
+        fill: #0f62fe;
```
should be using carbon tokens

davidicus commented on src/components/ProgressIndicator/_progress-indicator.scss:
```diff
@@ -1,82 +1,306 @@
+        fill: #0f62fe;
+
+        text {
+          fill: white;
```
should be using carbon tokens

davidicus commented on src/components/ProgressIndicator/_progress-indicator.scss:
```diff
@@ -1,82 +1,306 @@
-    flex: unset;
+      .sub-step {
+        svg {
+          fill: #0f62fe;
```
should be using carbon tokens

davidicus commented on src/components/ProgressIndicator/_progress-indicator.scss:
```diff
@@ -1,82 +1,306 @@
-.#{$prefix}--skeleton {
-  .#{$prefix}--progress-step {
-    flex: unset;
+      .sub-step {
```
same comment on class naming convention. Should follow BEM as outlined in Carbon repo

davidicus commented on src/components/ProgressIndicator/_progress-indicator.scss:
```diff
@@ -1,82 +1,306 @@
-        text-align: right;
+  .step-complete {
+    svg {
+      fill: #0f62fe;
```
should be using carbon tokens

davidicus commented on src/components/ProgressIndicator/_progress-indicator.scss:
```diff
@@ -1,82 +1,306 @@
+
+  .step-incomplete {
+    .main-step {
+      fill: #171717;
```
should be using carbon tokens

davidicus commented on src/components/ProgressIndicator/_progress-indicator.scss:
```diff
@@ -1,82 +1,306 @@
+
+    .sub-step {
+      svg {
+        color: #171717;
```
should be using carbon tokens

davidicus commented on src/components/ProgressIndicator/_progress-indicator.scss:
```diff
@@ -1,82 +1,306 @@
+  .step-disabled {
+    &:not(.step-invalid) {
+      .main-step {
+        fill: #bebebe;
```
should be using carbon tokens

davidicus commented on src/components/ProgressIndicator/_progress-indicator.scss:
```diff
@@ -1,82 +1,306 @@
+
+      .sub-step {
+        svg {
+          fill: #bebebe;
```
should be using carbon tokens

davidicus commented on src/components/ProgressIndicator/_progress-indicator.scss:
```diff
@@ -1,82 +1,306 @@
-      margin-left: 0;
+    &.step-invalid {
+      svg {
+        fill: #bebebe;
```
should be using carbon tokens

davidicus commented on src/components/ProgressIndicator/_progress-indicator.scss:
```diff
@@ -1,82 +1,306 @@
+
+    .label {
+      @extend .label;
+      color: #c6c6c6;
```
should be using carbon tokens

davidicus commented on src/components/ProgressIndicator/_progress-indicator.scss:
```diff
@@ -1,82 +1,306 @@
+    .label {
+      @extend .label;
+      color: #c6c6c6;
+      cursor: default !important;
```
we should not have `!important`. If higher specificity is needed it should be gradually raised through higher specificity .

davidicus commented on src/components/ProgressIndicator/_progress-indicator.scss:
```diff
@@ -1,82 +1,306 @@
+
+  .step-invalid {
+    svg {
+      fill: #dd3b41;
```
should be using carbon tokens

davidicus commented on src/components/ProgressIndicator/_progress-indicator.scss:
```diff
@@ -1,82 +1,306 @@
+
+    .line {
+      left: 0;
+      right: 24px;
```
no pixel values

davidicus commented on src/components/ProgressIndicator/_progress-indicator.scss:
```diff
@@ -1,82 +1,306 @@
+
+      &-sub {
+        left: 0;
+        right: 16px;
```
no pixel values

davidicus commented on src/components/ProgressIndicator/_progress-indicator.scss:
```diff
@@ -1,82 +1,306 @@
+
+  .label {
+    position: relative;
+    font-family: $IBMPlexSans;
+    font-size: 14px;
+    color: #161616;
+    letter-spacing: 0.16px;
+    line-height: 18px;
```
the site is down so I can't confirm but there may be a carbon text mixin that we should be using here.

davidicus commented on src/components/ProgressIndicator/ProgressIndicator.jsx:
```diff
@@ -1,141 +1,362 @@
+
+  const StepLine = () => {
+    const classes = classnames({
+      [`line`]: !complete && !subStep,
```
classes will all need a prefix and should follow the BEM syntax as outlined in Carbon repo

davidicus commented on src/components/ProgressIndicator/ProgressIndicator.jsx:
```diff
@@ -1,141 +1,362 @@
+
+  const StepLabel = () => {
+    const classes = classnames({
+      [`label`]: mainStep || subStep,
```
Same comment about the classes

davidicus commented on src/components/ProgressIndicator/ProgressIndicator.jsx:
```diff
@@ -1,141 +1,362 @@
+
+  const StepSecondaryLabel = () => {
+    const classes = classnames({
+      [`label-optional`]: mainStep || subStep,
```
same comment about the class naming convention

davidicus commented on src/components/ProgressIndicator/ProgressIndicator.jsx:
```diff
@@ -1,141 +1,362 @@
+  };
+
+  const classes = classnames({
+    [`step-current`]: current && !disabled,
```
prefixes and BEM styled class names

davidicus commented on src/components/ProgressIndicator/ProgressIndicator.jsx:
```diff
@@ -1,141 +1,362 @@
+  label: null,
+  secondaryLabel: null,
+  description: null,
+  onChange: null,
```
we should use our default function so component does not blow up if no `onChange` is provided.

tay1orjones replied:
Or nullcheck before calling the function https://github.com/IBM/carbon-addons-iot-react/issues/1401

tay1orjones commented on src/components/ProgressIndicator/ProgressIndicator.jsx:
```diff
@@ -1,141 +1,362 @@
-  );
+  useEffect(() => setCurrentStep(currentItemId), [currentItemId]);
+
+  const hasItems = () => newItems.length > 1;
+
+  const classes = classnames({
+    [`${iotPrefix}--progress-indicator`]: true,
+    [`${iotPrefix}--progress-indicator--vertical`]: isVerticalMode,
+    [className]: className,
+  });
+
+  return hasItems() ? (
```
`hasItems()` seems like an unnecessary abstraction

tay1orjones commented on src/utils/__tests__/__snapshots__/publicAPI.test.js.snap:
```diff
@@ -3227,13 +3239,19 @@ Map {
       "currentItemId": null,
+      "isClickable": false,
       "isVerticalMode": false,
-      "onClickItem": null,
```
Removing props is a breaking change. Can we leave it in, put a deprecation notice, and map it to a new prop?

tay1orjones commented on src/utils/__tests__/__snapshots__/publicAPI.test.js.snap:
```diff
@@ -3268,19 +3295,24 @@ Map {
-        "isRequired": true,
         "type": "arrayOf",
       },
-      "onClickItem": Object {
```
Same thing here on this prop - we're avoiding breaking changes right now

tay1orjones commented:
@lukefirth @dianatran18 this looks visually correct to me, but would you mind doing a quick review to ensure we look good here?

[Interactive stories for review can be found here](https://deploy-preview-1369--carbon-addons-iot-react.netlify.app/?path=/story/watson-iot-progressindicator--stateful)

rawa-resul commented:
> Should the focus and active states for adjoining steps be influencing one another? When I have an item selected and then hover over the next/previous, it increases the line below the initial item
> 
> ![Jul-14-2020 16-44-37](https://user-images.githubusercontent.com/3360588/87479622-6f1a3480-c5f1-11ea-856d-5dba58cf632d.gif)
> 
> @lukefirth @dianatran18 what's the desired focus/active states here with this interaction?

I have changed the focus to this:
![focus and hover](https://user-images.githubusercontent.com/61942902/87535154-ef42a780-c68e-11ea-8cbe-b590a6155aac.gif)

tay1orjones commented on src/components/ProgressIndicator/ProgressIndicator.jsx:
```diff
@@ -1,141 +1,364 @@
+    <ul
+      className={classes}
+      data-testid={`${iotPrefix}--progress-indicator-testid`}
+      onChange={onClickItem || handleChange}
```
To preserve the original signature of this we need to call `onClickItem` with the `onClickItem(items[index].id);`, correct?

rawa-resul replied:
No, by default the progress step onclick event will return the id instead of the index. Before we had a function called handleChange that would use items[index].id and pass it on to the callback function. But I have skipped that part as I am directly passing the id from the progress step.

tay1orjones commented on src/components/ProgressIndicator/ProgressIndicator.test.jsx:
```diff
@@ -1,22 +1,43 @@
+  { id: 'step5', label: 'Fifth Step' },
+];
 
-describe('ProgressIndicator', () => {
-  it('handleChange', () => {
-    const mockOnClickItem = jest.fn();
-    const wrapper = mount(<ProgressIndicator items={mockItems} onClickItem={mockOnClickItem} />);
-    // click the next step
-    wrapper.find('[tabIndex=0]').simulate('click');
-    expect(mockOnClickItem).toHaveBeenCalledWith('myid2');
-  });
-  it('if click item is not set it still works', () => {
-    const wrapper = mount(<ProgressIndicator items={mockItems} />);
-    // click the next step
-    wrapper.find('[tabIndex=0]').simulate('click');
-    expect(wrapper).toBeDefined();
-  });
```
I think this should be left in until we remove the `onClickItem `

rawa-resul replied:
I have added the handleChange test back although adjusted it to use screen now and the other test is actually being tested with one of the other ones I already had in there.

tay1orjones commented on src/components/ProgressIndicator/ProgressIndicator.jsx:
```diff
@@ -1,141 +1,364 @@
+  isVerticalMode: PropTypes.bool,
+  isClickable: PropTypes.bool,
+  setStep: PropTypes.func,
+  onClickItem: PropTypes.func,
```
I believe this prop is effectively deprecated, right? Could we add a deprecation warning to this component about this prop?

We have a helper for it:

https://github.com/IBM/carbon-addons-iot-react/blob/4d14293cc189abf9d222a11d0c5327cc347d0e00/src/components/PageTitleBar/PageTitleBar.jsx#L18-L21

rawa-resul replied:
I have added the onClickItem prop back. No need to deprecate it. I made some minor refactoring.

dianatran18 commented:
> > Should the focus and active states for adjoining steps be influencing one another? When I have an item selected and then hover over the next/previous, it increases the line below the initial item
> > ![Jul-14-2020 16-44-37](https://user-images.githubusercontent.com/3360588/87479622-6f1a3480-c5f1-11ea-856d-5dba58cf632d.gif)
> > @lukefirth @dianatran18 what's the desired focus/active states here with this interaction?
> 
> I have changed the focus to this:
> ![focus and hover](https://user-images.githubusercontent.com/61942902/87535154-ef42a780-c68e-11ea-8cbe-b590a6155aac.gif)

I double checked to see what the Carbon team was doing for their progress indicator. On hover, the text should turn the link (blue) color, and there should be a 1px underline. The focus state should be the same, but with a 3px underline.

The code here is correct: https://react.carbondesignsystem.com/?path=/story/progressindicator--default

<img width="1168" alt="Screen Shot 2020-07-21 at 9 24 36 AM" src="https://user-images.githubusercontent.com/43549567/88060594-7337dc00-cb34-11ea-9786-0a80a1df6af2.png">

rawa-resul commented:
> > > Should the focus and active states for adjoining steps be influencing one another? When I have an item selected and then hover over the next/previous, it increases the line below the initial item
> > > ![Jul-14-2020 16-44-37](https://user-images.githubusercontent.com/3360588/87479622-6f1a3480-c5f1-11ea-856d-5dba58cf632d.gif)
> > > @lukefirth @dianatran18 what's the desired focus/active states here with this interaction?
> > 
> > 
> > I have changed the focus to this:
> > ![focus and hover](https://user-images.githubusercontent.com/61942902/87535154-ef42a780-c68e-11ea-8cbe-b590a6155aac.gif)
> 
> I double checked to see what the Carbon team was doing for their progress indicator. On hover, the text should turn the link (blue) color, and there should be a 1px underline. The focus state should be the same, but with a 3px underline.
> 
> The focus and active states shouldn't be influencing each other either.
> 
> The code here is correct: https://react.carbondesignsystem.com/?path=/story/progressindicator--default
> 
> <img alt="Screen Shot 2020-07-21 at 9 24 36 AM" width="1168" src="https://user-images.githubusercontent.com/43549567/88060594-7337dc00-cb34-11ea-9786-0a80a1df6af2.png">

I refactored the css again and followed the original css file from carbon [here](https://github.com/carbon-design-system/carbon/blob/master/packages/components/src/components/progress-indicator/_progress-indicator.scss)

tay1orjones commented:
:tada: This PR is included in version 2.104.2 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/carbon-addons-iot-react)
- [GitHub release](https://github.com/carbon-design-system/carbon-addons-iot-react/releases/tag/v2.104.2)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
